modules/auxiliary/admin/vmware/vcenter_forge_saml_token.rb

```diff
@@ -0,0 +1,414 @@
+    'md' => 'urn:oasis:names:tc:SAML:2.0:metadata',
+    'xsi' => 'http://www.w3.org/2001/XMLSchema-instance',
+    'xs' => 'http://www.w3.org/2001/XMLSchema'
+  }.freeze
+
+  PREFIX_LIST = 'xsd xsi'.freeze
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'VMware vCenter Forge SAML Authentication Credentials',
+        'Description' => %q{
+          This module forges valid SAML credentials for vCenter server
+          using the vCenter SSO IdP certificate, IdP private key, and
+          VMCA certificates as input objects; you must also  provide
+          the vCenter SSO domain name and vCenter FQDN. The module will
+          return a session cookie for the /ui path that grants access to
+          the SSO domain as a vSphere administrator. The IdP trusted
+          certificate chain can be retrieved using Metasploit post
+          exploitation modules or extracted manually from
+          /storage/db/vmware-vmdir/data.mdb using binwalk.
+        },
+        'Author' => 'npm[at]cesium137.io',
+        'Platform' => [ 'linux' ],
+        'DisclosureDate' => '2022-04-20',
+        'SessionTypes' => [ 'meterpreter', 'shell' ],
+        'License' => MSF_LICENSE,
+        'References' => [
+          ['URL', 'https://www.horizon3.ai/compromising-vcenter-via-saml-certificates/']
+        ],
+        'Actions' => [
+          [
+            'Run',
+            {
+              'Description' => 'Generate vSphere session cookie'
+            }
+          ]
+        ],
+        'DefaultAction' => 'Run',
+        'DefaultOptions' => {
+          'USERNAME' => 'administrator',
+          'DOMAIN' => 'vsphere.local',
+          'RPORT' => 443,
+          'SSL' => true
+        },
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ IOC_IN_LOGS ]
+        },
+        'Privileged' => true
+      )
+    )
+
+    register_options([
+      OptString.new('USERNAME', [ true, 'The username to target using forged credentials', 'administrator' ]),
+      OptString.new('DOMAIN', [true, 'The target vSphere SSO domain', 'vsphere.local']),
+      OptString.new('VHOST', [true, 'DNS FQDN of the vCenter server']),
+      OptPath.new('VC_IDP_CERT', [ true, 'Path to the vCenter IdP certificate' ]),
+      OptPath.new('VC_IDP_KEY', [ true, 'Path to the vCenter IdP private key' ]),
+      OptPath.new('VC_VMCA_CERT', [ true, 'Path to the vCenter VMCA certificate' ])
+    ])
+
+    register_advanced_options([
+      OptInt.new('VC_IDP_TOKEN_BEFORE_SKEW', [ true, 'NOT_BEFORE seconds to subtract from current time, values 300 to 2592000', 2592000 ]),
+      OptInt.new('VC_IDP_TOKEN_AFTER_SKEW', [ true, 'NOT_AFTER seconds to add to current time, values 300 to 2592000', 2592000 ])
+    ])
+
+    deregister_options('Proxies')
+  end
+
+  def rhosts
+    datastore['RHOSTS']
+  end
+
+  def rport
+    datastore['RPORT']
+  end
```

Comment:
Please remove these. `rhosts` should not be accessed directly as modules are run for a single `rhost` value at a time. `rhost` & `rport` are already defined helper methods.